apsuite/commissioning_scripts/inj_traj_fitting.py

```diff
@@ -0,0 +1,230 @@
+
+from .base import BaseClass
+
+
+class Params:
+    """."""
+
+    def __init__(self):
+        """."""
+        self.simul_emitx = 3.5e-9
+        self.simul_emity = 3.5e-11
+        self.simul_espread = 1e-3
+        self.simul_bunlen = 5e-3
+        self.simul_npart = 1000
+        self.simul_cutoff = 6
+        self.count_rel_thres = 1/3
+
+
+class FitInjTraj(BaseClass):
+    """."""
+
+    def __init__(self, ring=None, sim_mod=None):
+        """."""
+        super().__init__(Params())
+        self.devices['sofb'] = SOFB(SOFB.DEVICES.SI)
+        self.model = ring if ring is not None else si.create_accelerator()
+        self.simul_model = sim_mod if sim_mod is not None else self.model[:]
+        self.model.vchamber_on = True
+        self.simul_model.vchamber_on = True
+
+        injp = pyaccel.lattice.find_indices(
+            self.model, 'fam_name', 'InjNLKckr')
+        self.model = pyaccel.lattice.shift(self.model, injp[0]+1)
+        self.simul_model = pyaccel.lattice.shift(self.simul_model, injp[0]+1)
+
+        self.famdata = si.get_family_data(self.model)
+        self.bpm_idx = np.array(self.famdata['BPM']['index']).flatten()
+        self.twiss, *_ = pyaccel.optics.calc_twiss(self.model)
+
+    def calc_traj(self, x0, xl0=0, y0=0, yl0=0, delta=0, size=160):
+        """."""
+        rin = np.array([x0, xl0, y0, yl0, delta, 0])
+        rout, *_ = pyaccel.tracking.linepass(
+            self.model, rin, self.bpm_idx[:size])
+        return rout[0, :], rout[2, :]
+
+    def calc_residue(self, vec, tx_meas, ty_meas):
+        """."""
+        tx_mod, ty_mod = self.calc_traj(*vec, size=tx_meas.size)
+        nanidcs = np.logical_or(np.isnan(tx_mod), np.isnan(tx_mod))
+        tx_mod[nanidcs] = 10.0
+        ty_mod[nanidcs] = 10.0
+        return np.hstack([tx_mod-tx_meas, ty_mod-ty_meas])
+
+    def calc_chisqr(self, residue):
+        """."""
+        return np.sum(residue*residue)/residue.size
+
+    def calc_jacobian(self, vec, size=160):
+        """."""
+        mat = np.zeros((2*size, 5))
+        zer = np.zeros(size)
+        dlt = 1e-5
+        for i in range(vec.size):
+            dvec = np.array(vec)
+            dvec[i] += dlt/2
+            res_pos = self.calc_residue(dvec, zer, zer)
+            dvec[i] -= dlt
+            res_neg = self.calc_residue(dvec, zer, zer)
+            mat[:, i] = (res_pos - res_neg)/dlt
+        return mat
+
+    def calc_init_vals(self, trajx, trajy):
+        """."""
+        etax_ave = np.mean(self.twiss.etax[self.bpm_idx])
+        x_ini, y_ini, xl_ini, yl_ini = trajx[0], trajy[0], 0, 0
+        de_ini = np.mean(trajx) / etax_ave
+        return np.array([x_ini, xl_ini, y_ini, yl_ini, de_ini])
+
+    def do_fitting(self, trajx, trajy, vec0=None, max_iter=5, tol=1e-6):
+        """."""
+        vec0 = vec0 if vec0 is not None else self.calc_init_vals(trajx, trajy)
+        res0 = self.calc_residue(vec0, trajx, trajy)
+        chi0 = self.calc_chisqr(res0)
+        vecs = [vec0, ]
```

Comment:
I thought about this, but this whole correction loop takes 1s to run and the visual inspection for the fitting quality is very direct. I decided to drop this idea because I think the importance of this data won't last long. Besides, with the resulting fitted vectors it would be easy to get `chi`in case the user wants it.

I could meet you half way and add a `full` flag to the input of the method in case the user wants the more detailed results. What do you think @murilobalves?

Reply:
Ok! The `chis` can be obtained from `vecs`, indeed. Since the correction loop is that fast, I think it's fine the way it's. Then, nevermind my comment.

Reply:
done!